File: leetcode/2. problem/java/97.交错字符串.java

/*
 * @lc app=leetcode.cn id=97 lang=java
 *
 * [97] 交错字符串
 */

// @lc code=start
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        // 动态规划
        // return extracted(s1, s2, s3);

        // test
        return test(s1, s2, s3);
    }

    private boolean test(String s1, String s2, String s3) {
        int len1 = s1.length(), len2 = s2.length(), len3 = s3.length();
        if (len1 + len2 != len3) {
            return false;
        }
        boolean[][] dp = new boolean[len1 + 1][len2 + 1];
        dp[0][0] = true;
        for (int i = 0; i <= len1; i++) {
            for (int j = 0; j <= len2; j++) {
                int idx = i + j - 1;
                if (i > 0) {
                    dp[i][j] = dp[i][j] || (dp[i - 1][j] && s1.charAt(i - 1) == s3.charAt(idx));
                }
                if (j > 0) {
                    dp[i][j] = dp[i][j] || (dp[i][j - 1] && s2.charAt(j - 1) == s3.charAt(idx));
                }
            }
        }
        return dp[len1][len2];
    }

    /**
     * 106/106 cases passed (3 ms)
     * Your runtime beats 67.41 % of java submissions
     * Your memory usage beats 92.16 % of java submissions (39.4 MB)
     * 
     * @param s1
     * @param s2
     * @param s3
     * @return
     */
    private boolean extracted(String s1, String s2, String s3) {
        // dp[i][j] = true, 表示s3的[i+j-1],可以是S1的[i]与s2[j]之前的字符组成
        int len1 = s1.length();
        int len2 = s2.length();
        int len3 = s3.length();
        if (len1 + len2 != len3) {
            return false;
        }
        boolean[][] dp = new boolean[len1 + 1][len2 + 1];
        dp[0][0] = true;
        for (int i = 0; i <= len1; i++) {
            for (int j = 0; j <= len2; j++) {
                int idx = i + j - 1;
                if (i > 0) {
                    dp[i][j] = dp[i][j] || (dp[i - 1][j] && s1.charAt(i - 1) == s3.charAt(idx));
                }
                if (j > 0) {
                    dp[i][j] = dp[i][j] || (dp[i][j - 1] && s2.charAt(j - 1) == s3.charAt(idx));
                }
            }
        }
        return dp[len1][len2];
    }
}
// @lc code=end

File: leetcode/2. problem/97.交错字符串.go
/*
 * @lc app=leetcode.cn id=97 lang=golang
 *
 * [97] 交错字符串
 */
package main

// @lc code=start
func isInterleave(s1 string, s2 string, s3 string) bool {
	// 创建 dp 数组
	return newFunction7(s1, s2, s3)

}

// 106/106 cases passed (0 ms)
// Your runtime beats 100 % of golang submissions
// Your memory usage beats 71.02 % of golang submissions (2 MB)
func newFunction7(s1 string, s2 string, s3 string) bool {
	len1, len2, len3 := len(s1), len(s2), len(s3)
	if len1+len2 != len3 {
		return false
	}

	dp := make([][]bool, len1+1)
	for i := 0; i < len1+1; i++ {
		dp[i] = make([]bool, len2+1)
	}
	dp[0][0] = true
	for i := 0; i <= len1; i++ {
		for j := 0; j <= len2; j++ {
			idx := i + j - 1
			if i > 0 {
				dp[i][j] = dp[i][j] || (dp[i-1][j] && s1[i-1] == s3[idx])
			}
			if j > 0 {
				dp[i][j] = dp[i][j] || (dp[i][j-1] && s2[j-1] == s3[idx])
			}
		}
	}
	return dp[len1][len2]
}

// @lc code=end
